fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java

```diff
@@ -184,20 +184,65 @@ public boolean canUsePipeline() {
      * Assign ParallelExecNum by PARALLEL_FRAGMENT_EXEC_INSTANCE_NUM in SessionVariable for synchronous request
      * Assign ParallelExecNum by default value for Asynchronous request
      */
-    public void setParallelExecNumIfExists() {
+    private void setParallelExecNumIfExists() {
         if (ConnectContext.get() != null) {
-            int pipelineDop = ConnectContext.get().getSessionVariable().getPipelineDop();
             int instanceNum = ConnectContext.get().getSessionVariable().getParallelExecInstanceNum();
-            int degreeOfParallelism = ConnectContext.get().getSessionVariable().getDegreeOfParallelism();
 
             if (ConnectContext.get().getSessionVariable().isEnablePipelineEngine()) {
-                parallelExecNum = pipelineDop > 0 ? instanceNum : degreeOfParallelism;
+                // In pipeline engine, we prefer inter-pipeline parallelism to inter-fragment parallelism
+                // So in default case, instanceNum should be 1, and DOP should be cores/2
+                int pipelineDop = ConnectContext.get().getSessionVariable().getPipelineDop();
+                if (pipelineDop > 0) {
+                    this.parallelExecNum = instanceNum;
+                    this.pipelineDop = pipelineDop;
+                } else {
+                    this.parallelExecNum = 1;
+                    this.pipelineDop = BackendCoreStat.getDefaultDOP();
+                }
             } else {
-                parallelExecNum = instanceNum;
+                this.parallelExecNum = instanceNum;
+                this.pipelineDop = 1;
             }
         }
     }
 
+    /**
+     * Adapt dop according to number of backends and number of instances
+     *
+     * @param numBackends  total number of backends
+     * @param numInstances total number of fragment instances
+     */
+    public void adaptPipelineDop(int numBackends, int numInstances) {
+        Preconditions.checkState(ConnectContext.get() != null &&
+                ConnectContext.get().getSessionVariable().isPipelineDopAdaptionEnabled() &&
+                getPlanRoot().canUsePipeLine());
+
+        int degreeOfParallelism = ConnectContext.get().getSessionVariable().getDegreeOfParallelism();
+        this.pipelineDop = Math.max(1, degreeOfParallelism / Math.max(1, numInstances / Math.max(1, numBackends)));
```

Comment:
in non-pipeline, if parallel_exec_fragment_instance_num = n,  fragment instance num of the PlanFragment depends on leftmost node, for ScanNode, the total degree of instance parallelism  is  numBackends x parallelExecNum, would be better to use this as a total dop in pipeline?